### PR TITLE
[Data] Image Classification Batch Inference with Ray Data

### PR DESCRIPTION
**Pull request by amogkam**
Updates @woshiyyya's image classification batch inference example using Ray Data APIs. 

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb
Comment (bveeramani) [inline doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: > The first parameter of map and map_batches is the user-defined function (UDF), which can either be a function or a class. Function-based UDFs will run as short-running [Ray tasks](https://docs.ray.io/en/latest/ray-core/key-concepts.html#tasks), and class-based UDFs will run as long-running [Ray actors](https://docs.ray.io/en/latest/ray-core/key-concepts.html#actors). For class-based UDFs, we use the compute argument to specify {class}ActorPoolStrategy <ray.data.dataset_internal.compute.ActorPoolStrategy> with the number of parallel actors. **And the batch_size argument indicates the number of images in each batch.**

This paragraph contrasts actor-based and task-based UDFs, so this last sentence seems out of place.

Comment (bveeramani) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: Also, while I feel like this is generally useful information, I don't think task-based UDFs are relevant to model inference.

Comment (bveeramani) [inline doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: > All of the images in the batch have been correctly classified as "tench" which is a type of fish.

Why are all the images tench? I'd expect to see a mix of different classes.

Comment (bveeramani) [inline doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]:
```python
def preprocess_images(image: Dict[str, Any]):
    return {"original_image": image["image"], 
            "transformed_image": transform(image["image"]),
           }
```

Would `row` or `record` be a better parameter name? I feel like `image["image"]` is kinda confusing.

Comment (bveeramani) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: Also, would `preprocess_image` be a better function name? This function works on individual images and not batches of images.

Comment (bveeramani) [inline doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: > for image, prediction in zip(prediction_batch["original_image"], prediction_batch["predicted_labels"]):

Super nit, but we should be consistent about our use of plural. Like, either use `"original_images"` and `"predicted_labels"`, or `"original_image"` and `"predicted_label"`. My preference is for singular.

**Review by bveeramani — APPROVED** [review]
Content overall LGTM. Will address stylistic/editorial changes in a follow-up PR.

File: doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb
Comment (amogkam) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: this dataset contains a lot of tench images :)

[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: updated these cells. Our formatting script does run on the notebook, but not sure how well it works.

[the remark this answers is not in the document]
Comment (amogkam) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: Added reference. But looks like UDF has been removed from the glossary.

Comment (amogkam) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: Updated to remove the task information.

Comment (amogkam) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: Updated both, good point.

Comment (amogkam) [reply doc/source/data/examples/pytorch_resnet_batch_prediction.ipynb:1]: updated!

**Review by matthewdeng — APPROVED** [review]
<div><br/>via <a href="https://giphy.com/disneyzootopia/">Disney Zootopia</a> on <a href="https://giphy.com/gifs/disneyzootopia-l2JHVUriDGEtWOx0c">GIPHY</a></div>

**Review by raulchen — APPROVED** [review]
> # Remove the `.cuda()` if using CPU.

We can check `torch.cuda.is_available()` to make the code more portable. 

> The batch_size argument indicates the number of images in each batch. The num_gpus argument specifies the number of GPUs needed for each ResnetModel instance. In this case, we want 1 GPU for each model replica.

Here better say "change `num_gpus=1` to `num_cpus=1`  if using CPUs for inference."

File: doc/source/data/examples/batch_inference_object_detection.ipynb [hunk 1]
```diff
@@ -912,12 +912,13 @@
...
-    "The first parameter of `map` and `map_batches` is the user-defined function (UDF), which can either be a function or a class. Function-based UDFs will run as short-running [Ray tasks](https://docs.ray.io/en/latest/ray-core/key-concepts.html#tasks), and class-based UDFs will run as long-running [Ray actors](https://docs.ray.io/en/latest/ray-core/key-concepts.html#actors). For class-based UDFs, we use the `compute` argument to specify the number of parallel actors. And the `batch_size` argument indicates the number of images in each batch.\n",
+    "The first parameter of `map` and `map_batches` is the user-defined function (UDF), which can either be a function or a class. Function-based UDFs will run as short-running [Ray tasks](https://docs.ray.io/en/latest/ray-core/key-concepts.html#tasks), and class-based UDFs will run as long-running [Ray actors](https://docs.ray.io/en/latest/ray-core/key-concepts.html#actors). For class-based UDFs, we use tthe `compute` argument to specify {class}`ActorPoolStrategy <ray.data.dataset_internal.compute.ActorPoolStrategy>` with the number of parallel actors. And the `batch_size` argument indicates the number of images in each batch.\n",
```
Comment (angelinalg) [inline doc/source/data/examples/batch_inference_object_detection.ipynb:921]:
```suggestion
    "The first parameter of `map` and `map_batches` is the user-defined function (UDF), which can either be a function or a class. Function-based UDFs will run as short-running [Ray tasks](https://docs.ray.io/en/latest/ray-core/key-concepts.html#tasks), and class-based UDFs run as long-running [Ray actors](https://docs.ray.io/en/latest/ray-core/key-concepts.html#actors). For class-based UDFs, we use the `compute` argument to specify {class}`ActorPoolStrategy <ray.data.dataset_internal.compute.ActorPoolStrategy>` with the number of parallel actors. And the `batch_size` argument indicates the number of images in each batch.\n",
```

File: doc/source/data/examples/index.rst [hunk 2]
```diff
@@ -9,13 +9,32 @@ Examples
...
 .. _data-recipes:
 
-Simple Data Processing Examples
--------------------------------
-
 Ray Data is a data processing engine that supports multiple data
 modalities and types. Here you will find a few end-to-end examples of some basic data
-processing with Ray Data on tabular data, text (coming soon!), and imagery (coming
-soon!).
+processing with Ray Data on tabular data, text (coming soon!), and images.
```
Comment (angelinalg) [inline doc/source/data/examples/index.rst:14]:
```suggestion
processing with Ray Data on tabular data, text (coming soon), and images.
```

**Review by angelinalg — APPROVED** [review]

**Pull request merged**